my-webxr-app/src/components/PCA.tsx

```diff
@@ -0,0 +1,175 @@
+        throw new Error("The input matrix must be symmetric.");
+    }
+    const evd = new EVD(covarianceMatrix, { assumeSymmetric: true });
+    const S = evd.realEigenvalues;
+    S.reverse();
+    return S;
+}
+
+/**
+ * Computes the eigenvectors from a covariance matrix.
+ * 
+ * It uses Eigenvalue Decomposition (EVD) to compute the eigenvectors. The eigenvectors are then sorted in descending order 
+ * of their corresponding eigenvalues, so that the first eigenvector corresponds to the direction of maximum variance in the data, 
+ * the second to the direction of second most variance, and so on.
+ * 
+ * @param covarianceMatrix - The covariance matrix to compute the eigenvectors from.
+ * @param checkSymmetry - Optional. Whether to check if the input matrix is symmetric. Default is true.
+ * @returns A matrix of eigenvectors (U).
+ * @throws {Error} If the input matrix is not symmetric and checkSymmetry is true.
+ */
+export function computeEigenvectorsFromCovarianceMatrix(covarianceMatrix: Matrix, checkSymmetry: boolean = true): Matrix {
+    // Lazy evaluation, checkSymmetry is false the rest of condition will not be evaluated, reducing computation.
+    if (checkSymmetry && !isSymmetricMatrix(covarianceMatrix)) {
+        throw new Error("The input matrix must be symmetric.");
+    }
+    const evd = new EVD(covarianceMatrix, { assumeSymmetric: true });
+    const U = evd.eigenvectorMatrix;
+    U.flipRows();
+    return U;
+}
+
+
+/**
+ * Performs Principal Component Analysis (PCA) on the input dataset.
+ * 
+ * The function first converts the input to a Matrix if it's not already one. Then, the dataset is standardized.
+ * The covariance matrix of the standardized dataset is calculated next.
+ * The eigenvectors of the covariance matrix are computed and used to transform the dataset into the principal component space.
+ * The transformed dataset is then truncated to only include the first kComponents principal components.
+ * 
+ * @param datasetMatrix - The dataset to perform PCA on. This can be a Matrix or a data structure that can be converted to a Matrix.
+ * @param kComponents - The number of principal components to keep.
+ * @param assert - Optional. Whether to enable assertions. Default is true. Turn off to reduce computations.
+ * @returns The transformed dataset, as a first principal kComponents  Matrix.
+ * @throws {Error} If the input dataset is not a Matrix or a 2D array and assert is true.
+ */
+export function computePCA(datasetMatrix: MaybeMatrix, kComponents: number, assert: boolean = true): Matrix {
+    datasetMatrix = convertToMatrix(datasetMatrix, assert);
+    datasetMatrix = standardizeDataset(datasetMatrix, assert);
+
+    const covarianceMatrix = calculateCovarianceMatrix(datasetMatrix, assert);
+    const U = computeEigenvectorsFromCovarianceMatrix(covarianceMatrix, assert);
+
+    const predictions = datasetMatrix.mmul(U);
+    return predictions.subMatrix(0, predictions.rows - 1, 0, kComponents - 1);
+}
```

Comment:
Do we need a way to choose which component we are reducing?

Reply:
Yes, thank you for bringing this up, this is where English fails on me. Let me know if I got anything wrong.

Given a data set of 1000 features (columns, dimensions, fields), we could produce 1000 PCA components. However, only a subset is significant, capturing the most variance in the data set. So I sort the eigenvectors corresponding to the eigenvalues descendingly, this way the 1st to k-th columns of the result PCA components matrix will be the most significant ones. So we could choose how many PCA components we want to see.

And if "choose which component we are reducing" means which fields on the original data set to perform PCA on, then I don't have that yet. It should be a simple columns array truncated and passed into the PCA function.

Reply:
Okay. I don't think I have a good enough grasp on PCA, so I'm going to do some more reading and get back to you.

Reply:
See the blog I attached to this pr, it may help. All of the functions here are the steps I break down from that blog.